presto-kafka/src/main/java/com/facebook/presto/kafka/KafkaSplitManager.java

```diff
@@ -100,10 +101,18 @@ public ConnectorPartitionResult getPartitions(ConnectorTableHandle tableHandle,
                         log.warn("No leader for partition %s/%s found!", metadata.topic(), part.partitionId());
                     }
                     else {
+                        // make sure the first broker is the partition leader
+                        List<HostAddress> isrList = part.isr().stream().map(KafkaSplitManager::brokerToHostAddress).collect(toList());
+                        HostAddress partitionLeader = HostAddress.fromParts(leader.host(), leader.port());
+                        if (!isrList.get(0).getHostText().equals(leader.host())
+                                || isrList.get(0).getPort() != leader.port()) {
+                            isrList.remove(partitionLeader);
+                            isrList.add(0, partitionLeader);
+                        }
                         builder.add(new KafkaPartition(metadata.topic(),
                                 part.partitionId(),
-                                HostAddress.fromParts(leader.host(), leader.port()),
-                                ImmutableList.copyOf(Lists.transform(part.isr(), KafkaSplitManager::brokerToHostAddress))));
+                                partitionLeader,
+                                ImmutableList.copyOf(isrList)));
```

Comment:
that is a bit late to create an immutable copy. I would probably unroll the whole process above by looping over what is returned by part.isr() and add depending on the condition to an immutable list builder.